### PR TITLE
[ONNX] Fix ignored small eps in layer normalization in fp16

### LINKED ISSUE #86580: [ONNX] small eps of LayerNorm will be exported as 0.0 in fp16 mode
### 🐛 Describe the bug

For `torch.nn.LayerNorm` in fp16 mode, when `eps` is smaller than `2^(-24)` (minimal fp16 positive number), it will be exported as a constant `0.0`. This is different from PyTorch behavior - it still keeps `eps` and necessary calculations as float in fp16 mode. Is it possible to fix this like adding casts in the beginning (to float) and at the end (back to fp16) of the [export function](https://github.com/pytorch/pytorch/blob/master/torch/onnx/symbolic_opset9.py#L2661-L2689) when running in fp16 mode, since some computations needed to be done in fp32? Thanks.

Minimal export case:
```py
import torch

class Net(torch.nn.Module):
    def __init__(self, C):
        super().__init__()
        self.layer_norm = torch.nn.LayerNorm(C, eps=1e-8)
    def forward(self, x):
        return self.layer_norm(x)

N, C = 8, 4
model = Net(C).cuda().half()
x = torch.randn(N, C).cuda().half()

torch.onnx.export(model, x, "test_layernorm_export_fp16.onnx", opset_version=12)
```

### Versions

```
Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.24.1
Libc version: glibc-2.31

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.0.0-1032-azure-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB
GPU 6: Tesla V100-SXM2-32GB
GPU 7: Tesla V100-SXM2-32GB

Nvidia driver version: 450.80.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.2
[pip3] pytorch-lightning==1.6.3
[pip3] torch==1.11.0
[pip3] torch-nebula==0.15.1
[pip3] torch-ort==1.11.0
[pip3] torchmetrics==0.7.1
[pip3] torchvision==0.12.0+cu113
[conda] magma-cuda113             2.5.2                         1    pytorch
[conda] mkl                       2021.4.0                 pypi_0    pypi
[conda] mkl-include               2021.4.0                 pypi_0    pypi
[conda] numpy                     1.22.2                   pypi_0    pypi
[conda] pytorch-lightning         1.6.3                    pypi_0    pypi
[conda] torch                     1.11.0                   pypi_0    pypi
[conda] torch-nebula              0.15.1                   pypi_0    pypi
[conda] torch-ort                 1.11.0                   pypi_0    pypi
[conda] torchmetrics              0.7.1                    pypi_0    pypi
[conda] torchvision               0.12.0+cu113             pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89869

Prior to this change, the symbolic_fn `layer_norm` (before ONNX version 17) always lose precision when eps is smaller than Float type, while PyTorch always take eps as Double. This PR adds `onnx::Cast` into eps related operations to prevent losing precision during the calculation.

### CODE REVIEW

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -2701,22 +2701,39 @@ def native_layer_norm(
 
     two_cst = symbolic_helper._generate_wrapped_number(g, 2.0)
     eps_cst = symbolic_helper._generate_wrapped_number(g, eps)
+    eps_dtype = _type_utils.JitScalarType.from_value(eps_cst)
 
     mean = g.op("ReduceMean", input, axes_i=axes)
     numerator = sub(g, input, mean)
+    # Cast it to eps dtype to avoid precision loss
+    cast_numerator = g.op(
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:2709]: Would it be necessary to cast all the time, or only when numerator is float16?

Comment (titaiwangms) [reply torch/onnx/symbolic_opset9.py:2709]: In torch, EPS is always Double, while in ONNX, we used to keep it Float32 all the time without any issue. This Cast would make us stay with Double all the time.

Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py:2709]: Since EPS is scalar I suspect torch would down cast it when computing, if it is following the same casting logic for arithmetic regarding python number and torch.tensor. You might want to check out `aten/src/ATen/native/cpu/layer_norm_kernel.cpp`

Comment (titaiwangms) [reply torch/onnx/symbolic_opset9.py]: Done

**Comment by titaiwangms** [comment]
@BowenBao PTAL

**Review by BowenBao — APPROVED** [review]
